qpid-jms-client/src/test/java/org/apache/qpid/jms/transports/netty/NettyTcpTransportTest.java

```diff
@@ -394,6 +399,123 @@ public void testZeroSizedSentNoErrors() throws Exception {
+            Transport nioTransport = createConnectedTransport(serverLocation, sharedTransportOptions);
+            try {
+                groupRef = getGroupRef(nioTransport);
+                assertNotNull(groupRef.group());
+            } finally {
+                nioTransport.close();
+            }
+
+            server.stop();
+
+            try {
+                groupRef.group();
+                fail("Should have thrown ISE due to being closed");
+            } catch (IllegalStateException expected) {
+                // Ignore
+            } catch (Throwable unexpected) {
+                fail("Should have thrown IllegalStateException");
+            }
+        }
+
+        assertTrue(!transportClosed);  // Normal shutdown does not trigger the event.
+        assertTrue(exceptions.isEmpty());
+        assertTrue(data.isEmpty());
+    }
+
+    @Test(timeout = 60 * 1000)
+    public void testSharedEventLoopGroups() throws Exception {
+        try (NettyEchoServer server = createEchoServer(createServerOptions())) {
+            server.start();
+
+            int port = server.getServerPort();
+            URI serverLocation = new URI("tcp://localhost:" + port);
+            final TransportOptions sharedTransportOptions = createClientOptions();
+            sharedTransportOptions.setUseKQueue(false);
+            sharedTransportOptions.setUseEpoll(false);
+            sharedTransportOptions.setSharedEventLoopThreads(1);
+            Transport sharedNioTransport1 = createConnectedTransport(serverLocation, sharedTransportOptions);
+            Transport sharedNioTransport2 = createConnectedTransport(serverLocation, sharedTransportOptions);
+            assertSame(getGroupRef(sharedNioTransport1).group(), getGroupRef(sharedNioTransport2).group());
+            final EventLoopGroup sharedGroup = getGroupRef(sharedNioTransport1).group();
+
+            sharedNioTransport1.close();
+            assertFalse(sharedGroup.isShutdown());
+            assertFalse(sharedGroup.isTerminated());
+            sharedNioTransport2.close();
+            assertTrue(sharedGroup.isShutdown());
+            assertTrue(sharedGroup.isTerminated());
+
+            server.stop();
+        }
+
+        assertTrue(!transportClosed);  // Normal shutdown does not trigger the event.
+        assertTrue(exceptions.isEmpty());
+        assertTrue(data.isEmpty());
+    }
+
+    @Test(timeout = 60 * 1000)
+    public void testUnsharedEventLoopGroups() throws Exception {
+        try (NettyEchoServer server = createEchoServer(createServerOptions())) {
+            server.start();
+
+            int port = server.getServerPort();
+            URI serverLocation = new URI("tcp://localhost:" + port);
+
+            final TransportOptions unsharedTransportOptions1 = createClientOptions();
+            unsharedTransportOptions1.setUseKQueue(false);
+            unsharedTransportOptions1.setUseEpoll(false);
+            unsharedTransportOptions1.setSharedEventLoopThreads(1);
```

Comment:
Why is the 'unsharedTransportOptions1' (and 2 below) asking for a shared group? Seems like this is not in fact "testUnsharedEventLoopGroups" but really "testSharedEventLoopGroupsOfDifferentSizes".

If the variable names and test name were updated to retain this test of behaviour, it should then also probably add another transport to verify there is in fact still any sharing amongst the same-size groups, which it doesnt currently check.

Then the other test would need added